middle_end/flambda2/terms/flambda_primitive.ml

```diff
@@ -1339,33 +1353,35 @@ let effects_and_coeffects_of_binary_primitive p : Effects_and_coeffects.t =
     if Flambda_features.float_const_prop ()
     then No_effects, No_coeffects, Strict
     else No_effects, Has_coeffects, Strict
+  | Check_alignment _ -> Only_generative_effects Mutable, Has_coeffects, Strict
```

Comment:
I'm not sure effects are the right way for this primitive. As I understand it the "problem" is that the GC can move around values, which can change the alignment of the start of the value, and thus the result of the `Check_alignment` primitive. The idea would then be to give the primitive some effects and coeffects to make it less likely to be moved across GC points that could change its alignment, is that correct ?

But even then, I'm not sure what is the use of these effects: indeed the result of `Check_alignment` can change across a call to the CG, but it's not really what is tracked by the effects and coeffects, and the actual property that one would like to have/ensure is that no GC occurs between the alignment check and the access, but unfortunately, I don't think we have a mechanism for that in flambda2 currently.

Reply:
After further discussion, we will only provide the safe alignment check on bigarrays for now.

Reply:
If we only check bigarrays, we probably don't need to register any effects. Maybe coeffects if we need to consider changes to the underlying array.

Reply:
Right - I don't think we need to consider changes to the underlying array either, since we only care about its address, right?

Reply:
There were discussions in the past about what C code should be allowed to do with bigarrays, and some people mentioned that they wanted to be able to replace the pointer in a bigarray with a null pointer (after freeing the previous pointer) so that memory could be reclaimed without waiting for a GC. I believe ocaml/ocaml#389 is relevant.
My opinion is that it shouldn't matter, and `No_coeffects` is fine.